Implementing Janet core

fuzza opened the pull request:
This PR is currently in development, so it's for review and discussions only

# What's done:

## Actions
Currently any object that conforms to Equatable protocol could be used as an action

## ActionState
Action State is and enumeration that state of action's execution. It could have one of four states:
`start`, `progress`, `success`, `error`. 
Each state holds an action and could contain some metadata specific for this state.

## ActionService protocol
It's a type-erased wrapper over specific Janet Service. 
Janet works with all services using this protocol independent from their types. It allows to trick Swift's type system and use single reactive pipeline for all actions.

## TypedActionService
Generic protocol that represents specific ActionService implementation. 
It has `associated type` that specifies exact action being processed by this service. 
There is an extension for TypedActionService that adds default implementation of ActionService's `send`, `cancel` and `acceptsAction` methods. 
Default implementation performs actions typecasting and provides a simple matching for supported action based on protocol's associatedtype.

## ActionHolder
Since service could contain some logic that needs to modify action in some manner, there is an ActionHolder struct that simply holds two actions:
- origin action is needed to determine whether ActionState returned from service is match to action that was sent
- modified action - exact action that was executed by service
When we are talking that Service works with *action* we actually mean *ActionHolder*

## ActionPair
Actual that is retrieved from server should contain ActionHolder to determine `which` action was processed and ActionState to determine `what` happened with this action. 
ActionPair is a type alias for tuple from these two elements. 
Services report about their execution by sending ActionPairs.

## ActionServiceCallback
Provides set of callback methods used by services to report about action results
It's reponsible for building correct ActionPairs and sending it to shared pipeline.

## SharedPipeline
Shared pipeline is a PublishSubject that aggregates results of execution for all actions in scope of single Janet instance.

## Janet
Main class that is responsible for mediation between actions and services.
It creates SharedPipeline, creates ActionServiceCallback with pipeline and injects it to all ActionServices passed to Janet instance via constructor.

# Interaction
Client code interacts with Janet by sending Actions and observing ActionState
Janet interacts with Services by sending ActionHolders and
Services interact with Janet by sending ActionPairs to SharedPipeline via ActionServiceCallback

fuzza committed:
Add action state enumeration + tests

fuzza committed:
Add action service callback

- Add ActionHolder struct to store both origin and modified actions
- Add ActionServiceCallback to report action state from services to pipeline
- Add unit tests

fuzza committed:
Add Equatable requirement to action

fuzza committed:
Minor changes in namings and access modifiers

SwiftyJanetCI commented:
<!--
  0 Errors
  2 Warnings: This PR seems to be big. Is th...
  2 Messages: CI job config is updated
  1 Markdown
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR seems to be big. Is there any chance to reduce it’s size to simplify and speedup code review?</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">The <code>Cartfile</code> or <code>Cartfile.resolved</code> was updated, but there were no changes in the <code>podspec</code>. Did you forget updating the <code>podspec</code>?</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">CI job config is updated</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Fastlane config is updated</td>
    </tr>
  </tbody>
</table>

### Slowest tests 

File | Name | Time |
| --- | ----- | ----- |
| SwiftyJanet_iOS_Tests.JanetTests | test_doCancel_sendsErrorToPipeline_andAssertsIfServiceNotFound | 0.011 |


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

fuzza committed:
Cache carthage artifacts on CI

fuzza committed:
Improve test tooling

- Integrate RxTest and Nimble
- Revert Test folder excluded from swiftlint

fuzza committed:
Implement send method in Janet

- Add send method and send action observable
- Change errors to assertions for doSend and doCancel

fuzza committed:
Fix swiftlint

demalex commented on Sources/ActionHolder.swift:
```diff
@@ -0,0 +1,24 @@
+  }
+  
+  func isOrigin(action: Any) -> Bool {
+    guard let castedAction = action as? A else {
```
Since there is down cast to `A`, and `A` is `Equatable`, maybe is more correct to pass action as `Equatable` instead of `Any`? Something like `func isOrigin<T: Equatable>(action: T) -> Bool`

fuzza replied:
In my mind It doesn't matter since we anyway cast `action` parameter to ActionHolder's generic type `A` that has Equatable requirement. If we decide to use some specific type requirement for `action` parameter, Tit might be useful to create a type alias for action type that currently is defined as Equatable to avoid semantical confusion.

 I mean something kinda
 `typealias JanetAction = Equatable`

What do you think, will it improve clarity?

demalex commented on Sources/ActionServiceCallback.swift:
```diff
@@ -0,0 +1,34 @@
+public final class ActionServiceCallback {
+  let pipeline: SharedPipeline
+  
+  init(pipeline: PublishSubject<Any>) {
```
I there any reason, why you define init pipeline type `PublishSubject<Any>` and have pipeline variable as `SharedPipeline`?

fuzza replied:
Since we need to push ActionPairs of different types to this pipeline (e.g. `ActionPair<String>`, `ActionPair<Int>` etc.) we need to make it type-erased. Unfortunately Swift type system doesn't allow to use protocols in generics, so Any is the only choice here that I can imagine.

demalex commented on Sources/ActionHolder.swift:
```diff
@@ -0,0 +1,24 @@
+import Foundation
+
+public struct ActionHolder <A: Equatable> {
```
Since it is `ActionHolder`, maybe it will bring more readability if generic type name `A` will be changed to `Action`?

fuzza replied:
Definitely makes sense to me. 
I think we should do it for all action generics across the project. 
The only place where I think brevity is more important, is a function's generic type, so I'd suggest to proceed use single characters for function generic names
